tensorrt_llm/_torch/pyexecutor/model_engine.py

```diff
@@ -1196,28 +1193,26 @@ def _prepare_tp_inputs(
                 ) if mrope_rotary_cos_sin.device == 'cpu' else mrope_rotary_cos_sin
                 mrope_config['mrope_rotary_cos_sin'].append(
                     mrope_rotary_cos_sin.to('cuda', non_blocking=True))
-            request.py_batch_idx = batch_idx
-            batch_idx += 1
+            request.py_batch_idx = request.seq_slot
 
-        num_ctx_requests = batch_idx
+        num_ctx_requests = len(scheduled_requests.context_requests)
         num_ctx_tokens = len(input_ids)
         new_tokens_device, new_tokens_lens_device, next_draft_tokens_device = None, None, None
         if new_tensors_device is not None:
             # speculative decoding cases: [batch, 1 + draft_len], others: [batch]
             new_tokens_device = new_tensors_device.new_tokens
-            if self.without_logits:
-                assert isinstance(new_tensors_device, SampleStateTensorsMTP)
-                new_tokens_lens_device = new_tensors_device.new_tokens_lens  # [batch]
-                next_draft_tokens_device = new_tensors_device.next_draft_tokens  # [batch, draft_len]
+        is_mtp = isinstance(new_tensors_device, SampleStateTensorsMTP)
```

Comment:
Shouldn't this code go inside the `if new_tensors_device is not None:` block?

Reply:
No, not necessarily, since `new_tokens_lens_device` and `next_draft_tokens_device` are required to exist if `isinstance(new_tensors_device, SampleStateTensorsMTP)`  - basically, this is a more explicit typeguard